### PR TITLE
Agent quality tracker

### PR DESCRIPTION
**Pull request by tbak**
### Description of the Change

To prevent scheduling tasks on agent nodes where there were recent failures,
we need to track all recent errors on each agent instance, and aggregate them into
a single number describing agents quality. This information is used by `AgentManagementFitnessCalculator` as a multiplier to adjust (lower), the computed fitness value.

### CODE REVIEW

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        }
+
+        PlacementHistory placement = tryGetOrCreatePlacementHistory(task);
+        if (placement != null) {
+            placement.onTaskFailure(errorKind);
+        }
+    }
+
+    private PlacementHistory tryGetOrCreatePlacementHistory(Task task) {
+        String hostId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_HOST);
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: this is the host name

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        }
+
+        return placement;
+    }
+
+    private boolean isStartedTask(Task task) {
+        return JobFunctions.findTaskStatus(task, TaskState.Started).isPresent();
+    }
+
+    private enum ErrorKind {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: add javadocs as to what each `ErrorKind` means

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                invariants.inconsistent("Task without agent id assigned: taskId=%s", task.getId());
+                return null;
+            }
+            placement = agentManagementService.findAgentInstance(instanceId).map(PlacementHistory::new).orElse(null);
+        }
+
+        return placement;
+    }
+
+    private boolean isStartedTask(Task task) {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: I have seen these types of utilities a few times. Maybe we should start moving them to `JobFunctions` or another utility class

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            return Math.max(0, Math.min(1, 1 - failureLevel));
+        }
+
+        private void onTaskStarted() {
+            errorTimeSeries.clear();
+        }
+
+        private void onTaskFailure(ErrorKind errorKind) {
+            double weight;
+            switch (errorKind) {
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: I think you should move each weight to a constant and provide a javadoc mentioning the reasoning that went into choosing the weight. For instance, why is `LocalStartingError` a higher weight than `LostAfterStarted`?

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/fitness/AgentManagementFitnessCalculator.java [hunk 5]
```diff
@@ -66,10 +70,16 @@ public double calculateFitness(TaskRequest taskRequest, VirtualMachineCurrentSta
             logger.debug("Ignoring instanceGroupId: {} because it was not found in agent management", instanceGroupId, e);
         }
         if (instanceGroup != null) {
+            double quality = Math.min(1.0, agentQualityTracker.qualityOf(targetVM.getHostname()));
+            if (quality <= 0) {
+                // If we have no information about the agent, we have to assume something.
+                quality = 0.5;
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/fitness/AgentManagementFitnessCalculator.java]: move to constant

File: titus-common/src/main/java/com/netflix/titus/common/util/TimeSeriesData.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    private final long stepMs;
+
+    private final BiFunction<Double, Long, Double> adjuster;
+    private final Clock clock;
+
+    private final Queue<Pair<Double, Long>> offers = new ConcurrentLinkedQueue<>();
+    private final ConcurrentLinkedDeque<Pair<Double, Long>> data = new ConcurrentLinkedDeque<>();
+    private final AtomicReference<Pair<Double, Long>> lastAggregate;
+
+    private final Pair<Double, Long> nothing;
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/util/TimeSeriesData.java]: +1

File: titus-common/src/main/java/com/netflix/titus/common/util/TimeSeriesData.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiFunction;
+
+import com.netflix.titus.common.util.time.Clock;
+import com.netflix.titus.common.util.tuple.Pair;
+
+/**
+ * Simple, concurrent time series data collector.
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/util/TimeSeriesData.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+    @PreDestroy
+    public void shutdown() {
+        ObservableExt.safeUnsubscribe(agentStreamSubscription, jobStreamSubscription);
+    }
+
+    @Override
+    public double qualityOf(String agentHostName) {
+        PlacementHistory history = agentsPlacementHistory.get(agentHostName);
+        if (history == null) {
+            return -1;
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 1 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 2 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 4 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: Done

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/fitness/AgentManagementFitnessCalculator.java (hunk 5 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/fitness/AgentManagementFitnessCalculator.java]: +1

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 9]
```diff
@@ -32,6 +32,38 @@
...
+     */
+    private static final double LOST_BEFORE_STARTED_WEIGHT = 0.3;
+
+    /**
+     * A container loss after it was started is less likely to be an agent issue. We give it the lowest score.
+     */
+    private static final double LOST_AFTER_STARTED_WEIGHT = 0.1;
+
+    /**
+     * A container starting error due to an agent local problems. We give it a highest score.
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: remove `an`

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 10]
```diff
@@ -32,6 +32,38 @@
...
+     */
+    private static final double LOCAL_STARTING_ERROR_WEIGHT = 0.5;
+
+    /**
+     * Not agent related issue us not counted as an error.
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: Issue not related to agent; not counted as error

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 11]
```diff
@@ -32,6 +32,38 @@
...
+     */
+    private static final double SYSTEM_STARTING_ERROR_WEIGHT = 0;
+
+    /**
+     * A container crash may indicate the underlying agent degradation. We give it a middle score value.
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: remove `the`

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 12]
```diff
@@ -193,15 +225,30 @@ private PlacementHistory tryGetOrCreatePlacementHistory(Task task) {
...
+         */
+        LostAfterStarted,
+
+        /**
+         * A task failed to start on an agent, due to the agent's local issues.
+         */
         LocalStartingError,
+
+        /**
+         * A task failed to start on an agent, due to external to the agent system issues (AWS API rate limiting, network partitioning, etc).
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: due to external to the agent system issues -> due to external system issues

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java [hunk 13]
```diff
@@ -193,15 +225,30 @@ private PlacementHistory tryGetOrCreatePlacementHistory(Task task) {
...
+         */
         SystemStartingError,
+
+        /**
+         * A running container crashed due to an agent internal issues.
```
Comment (corindwyer) [inline titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: remove `an`

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 9 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: Fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 10 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: Fixed

File: titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java (hunk 11 above)
Comment (tbak) [reply titus-server-master/src/main/java/com/netflix/titus/master/scheduler/ContainerFailureBasedAgentQualityTracker.java]: Fixed

**Pull request merged**